### DISCUSSION TITLE
Confused with clustered index

### DISCUSSION THREAD

**Discussion by Derek-X-Wang** [tywalch/electrodb#193] [Q&A] [answered]
Thanks for this awesome library.

I'm trying to decide if I shall use `clustered` or `isolated` index.

As the document said, the `clustered` index can leverage the sort key when querying but will bring cost if trying to fetch the individual item. And it has a different key storage formats from the `isolated` index.

Could you give some examples of how the key will look with `clustered` or with `isolated`?

`repository` collection with `clustered` index: [link](https://electrodb.fun/?#code/PQKgBAsg9gJgpgGzARwK5wE4Es4GcA0YuccYGeqCALgUQBYCG5YA7llXWAGbZwB2MXGBDAAUKKwBbAA5QMVMAG8wAUT5V2AT0IBlTADcsAY1IBfbhiiSwAIkRwjVSzABGNgNzijUPrgVUGFwRSAF5bTShUDAB9AKC4aL4GSTgPcVAwABUGXABrITUNKk1hMW9fBXJZXHY5HCEwvjgWVXUtAApRMCUu7rBJWEQALh6+vv4izRGbACU4atqMTRt8XrH9TBqfaYBGFbW+4gxDE2nsvIBBaWkbA9NVsYYqJywXVCo8EcUD7qqoGqock0AEkYF8fn1itI4NM-Ng+ABzfZjMbkNBYchgsBOdAQ+4QorBcEoyGaaGwl6I5EkshwdGYkY4uB4h4o+C4IzYaQabajGlQmG2OFYKmsklo1AYuBYpksgkMBG4Yn8smCmzEKjUknsOCSJVCylIuUkzlwJ7Si5UZUkgXTPioSQuTBatlwLgMShWsDtACUYBCAD4wAARc0AOj4UBYvrFKOAwDARgYfEjCid-VgWC4OBgYAYXA+GET5HNMAhvzNMAA8nwEFNsRhcST8STUNIYKXLdaUbbbPbHc7Y2N46wnkZOFw5InGIi8NioHm+CUni83h9y6OqOPpiAXcOExrE1FyOpsVI8AEZKw6PwwG2Ox8yzSiHAvb7-UHQx8I1GYxuS9Wtb1rKzZ3LGIrwAAHp8fKovM-yLPW3zPt4CDBI4WC8jYfwAkCe6kuSthGAgqB+Jg0r4d00i5N2JLZogWI2NRlF9N4MgIR8IwANrYfBuFLKCNgALobi2NJ5LRKL0QgjF5Cx3Qjqa5qJlYCwfGAuRwJoQiTkW0hMBoDBILgcgKJpJRoJg9Qbt0bFqYKXEic+pjGn0LndGJyhxMEYAuT6niiBkAAyIr5K0kylKI5R+GACChQ0YBNC0hQdL0yF9AM8AIJJExaNMABqWDwFAIV8Lk+EbBgWx8Ls+FHCcarnLkDDXLcYxid0K7YGuMHpWMcVlaCkndL26qGixEpSjKjbMs2Q60mpQJDbBPaqhS8JIvNFb0tKjIza5YwwFYDAisN2JrQaG0TXSkoMg2TYoh1fRRNlK1jKNwqiv+N1TXtD3tfNhKCn1q2EWNV3zU93TspyWDcphNVvQRaqfZtB2HN45D6iD70XeDIpo3N8qKmdH2vixOp6utBNtY981KY+XZIyNeP9k6GAsfA7qeiM76BiG4aRtGPpbWAinJqmYDpplWY5nmBaYMWZqPt9DCAXWf2zXTEL3p2Xo48jdoOuzLEjiwY4TlO47Jgic6Aouy7PN17xayS5tbnQO6mwer5HhgJ7+OefjJNI163rrKvPhqvN+vzX5wD+wuiwBNYa-drtuWBBwQXA0HYxCOGIWdqHoTyiO8YtSwsaNxGkYWFGi9RZ3dNJjHMaLtmqRxDkVxxS0wMJomixJzN9K3sLlR3YsJgzpB2d3GlaTpU76fIWBGUQpmLxZ6C8LgNkqexAI9xych4CstgDbkglOTS7na+1vSedigQ+X5AXRQoLWh40zRgHoxxjBwHaMoQugI96ECvkIUw-l0jgAAOrYHUpZPe84wC2yaBgZSq9kivk2GAHw2IbxkCwAiOgCgRBeB8DFD4yR-S2AROQfgaRP53iOPQmwMAwx0FQJoPgdAWHUIUNISwAArBwCgwg2ExqkTwrCAh5A4QAFgAKwAFoABsGiABMAAGQRFQFp9wErmKRqjNE6P0QFb+YYS4SIRvvboYYwFAhAUY-iIIYCEFPljaYABOfRvkfS9DDAiKAvpxA2NyhoPAzi+KLHqKE18biXEmKCaE8JwTRCpMSbPEBvRUmeNhB2XA1IgbFOpDDLkZdim4AHqIGBGTfQBSgbYksHx8ndCvstJiURpDBGpIUnpVAwzmwQOOSpx1TpCn4JhDmrIXrTDGRM1kPiYI2H0YQBpPommwNEEAA)
`repository` collection with `isolated` index: [link](https://electrodb.fun/?#code/PQKgBAsg9gJgpgGzARwK5wE4Es4GcA0YuccYGeqCALgUQBYCG5YA7llXWAGbZwB2MXGBDAAUKKwBbAA5QMVMAG8wAUT5V2AT0IBlTADcsAY1IBfbhiiSwAIkRwjVSzABGNgNzijUPrgVUGFwRSAF5bTShUDAB9AKC4aL4GSTgPcVAwABUGXABrITUNKk1hMW9fBXJZXHY5HCEwvjgWVXUtAApRMCUu7rBJWEQALh6+vv4izRGbACU4atqMTRt8XrH9TBqfaYBGFbW+4gxDE2nsvIBBaWkbA9NVsYYqJywXVCo8EcUD7qqoGqock0AEkYF8fn1itI4NM-Ng+ABzfZjMbkNBYchgsBOdAQ+4QorBcEoyGaaGwl6I5EkshwdGYkY4uB4h4o+C4IzYaQabajGlQmG2OFYKmsklo1AYuBYpksgkMBG4Yn8smCmzEKjUknsOCSJVCylIuUkzlwJ7Si5UZUkgXTPioSQuTBatlwLgMShWsDtACUYBCAD4wAARc0AOj4UBYvrFKOAwDARgYfEjCid-VgWC4OBgYAYXA+GET5HNMAhvzNMAA8nwEFNsRhcST8STUNIYKXLdaUbbbPbHc7Y2N46wnkZOFw5InGIi8NioHm+CUni83h9y6OqOPpiAXcOExrE1FyOpsVI8AEZKw6PwwG2Ox8yzSiHAvb7-UHQx8I1GYxuS9Wtb1rKzZ3LGIrwAAHp8fKovM-yLPW3zPt4CDBI4WC8jYfwAkCe59NIuTdiS2aIFiNiEfhYzeDICEfCMADa2HwbhSygjYAC6G4tjSeTESipEIOReRUX0I6muaiZWAsHxgLkcCaEIk5FtITAaAwSC4HICjySUaCYPUG7dDRMmCgxXHPqYxp9FZ3Q8cocTBGAVk+p4ogZAAMiK+StJMpSiOUfhgAg3kNGATQtIUHS9MhfQDPACD8RMWjTAAalg8BQF5fC5PhGwYFsfC7PhRwnGq5y5Aw1y3GMPHdCu2BrjBsVjCFOWgvx3S9uqhqiRKUoyo2zLNkOtIyUCHWwT2qoUvCSKjRW9LSoyQ3WWMMBWAwIqddiM0GnNfV0pKDINk2KJ1X0USJVNYzdcKor-kdA0rWdtWjYSgotdN5L7SK81rX07Kclg3KYUVN2kj9PUHaNF3dBycjNRud29bD70KvqX23Xt6qvqJOp6rNf01edo0SY+XYQ11OP9k6GCifA7qeiM76BiG4aRtGPoLWA4nJqmYDpvFWY5nmBaYMWZqPo9DCAXWL3DaTEL3p2XpY5Daq04OG4jiwY4TlO47Jgic6Aouy7PI17yKySetbnQO6iSOh5GMeExnikfjJNI163ir0vPhqLN+mzX5wD+XM8wBNby6dNs2WBBwQXA0GYxCOGITtqHoTy4PMeNSyiYRO3dIJ5GUTzxnSXRZn53RE0wJx3E83xVN9GXsK5ZXvMJuTpAmTXckKUpU6qfIWAaUQ2lD3p6C8LgRlSbRAK1wj5C4Csthtbk7EWTStlK7VvT2digROS5bmBQoVU+40zRgHoxzGHA7TKBngLz4Q29CKYrnpOAAB1bAsl9Lz3nGAE2TQMCSTHskV8mwwA+GxDeMgWAER0AUCILwPggofGSP6WwCJyD8DSFfO8RwCE2BgGGOgqBNB8DoKQnBChpCWAAFYOAUGEGw3hyBMIqKfPIlCAAsABWAAtAANkkQAJgAAz8KCu-BuIiJHSPkWkUQN8wzZ04WDBe3QwzKKWK-Ma9c2K5l-r0MMCIoC+nENo5KGg8BGJYoseoNjXymOMSCSxPobF2J9OIYxHi+6v16D4ya6oOwb1ZB9WEjdWRAy5LnBJuBG6iF-gE30blv46JLB8cJ3Rt5ROkFEaQwRqSRPIlQMMesEDjmpBtSQW087ED4JhemrIrrTHqY01ka8YI2AUYQTJ-jbE5PEEAA)

The queries look the same for both `clustered` and `isolated`.  Is this expected?
```
// clustered index, scores is sort key
app.collections.repository({ repositoryId, scores: "90" }).go()
// isolated index
app.collections.repository({ repositoryId}).go()
// The result in playground looks the same
```

Thanks

**Comment by tywalch** [comment, marked as answer]
I @Derek-X-Wang 👋 

Index types ultimately impact how the sort key is structured, which impacts how the values are stored in DynamoDB. In DynamoDB, items stored in the order of how their sort key _sorts_ (duh, I know haha). As a result though, if you're trying to perform a partial query to retrieve multiple entities (a "collection" of entities), the results you get returned to you come back in this order as well. This isn't a problem if the items under your partition key can be retrieved in a single query -- however if your partition key has many items, and represent many entities, then it becomes important to consider how your items sort within your partition key. Should they be grouped together or mixed in with one another.

The effect of index type becomes visible in the sort key when you at least one sort key composite attribute. Here's what I mean using the examples you provided:

`get` a `link` with a `clustered` index: [link](https://electrodb.fun/?ssl=144&ssc=69&pln=142&pc=1#code/PQKgBAsg9gJgpgGzARwK5wE4Es4GcA0YuccYGeqCALgUQBYCG5YA7llXWAGbZwB2MXGBDAAUKKwBbAA5QMVMAG8wAUT5V2AT0IBlTADcsAY1IBfbhiiSwAIkRwjVSzABGNgNzijUPrgVUGFwRSAF5bTShUDAB9AKC4aL4GSTgPcVAwABUGXABrITUNKk1hMW9fBXJZXHY5HCEwvjgWVXUtAApRMCUu7rBJWEQALh6+vv4izRGbACU4atqMTRt8XrH9TBqfaYBGFbW+4gxDE2nsvIBBaWkbA9NVsYYqJywXVCo8EcUD7qqoGqock0AEkYF8fn1itI4NM-Ng+ABzfZjMbkNBYchgsBOdAQ+4QorBcEoyGaaGwl6I5EkshwdGYkY4uB4h4o+C4IzYaQabajGlQmG2OFYKmsklo1AYuBYpksgkMBG4Yn8smCmzEKjUknsOCSJVCylIuUkzlwJ7Si5UZUkgXTPioSQuTBatlwLgMShWsDtACUYBCAD4wAARc0AOj4UBYvrFKOAwDARgYfEjCid-VgWC4OBgYAYXA+GET5HNMAhvzNMAA8nwEFNsRhcST8STUNIYKXLdaUbbbPbHc7Y2N46wnkZOFw5InGIi8NioHm+CUni83h9y6OqOPpiAXcOExrE1FyOpsVI8AEZKw6PwwG2Ox8yzSiHAvb7-UHQx8I1GYxuS9Wtb1rKzZ3LGIrwAAHp8fKovM-yLPW3zPt4CDBI4WC8jYfwAkCe6kuSthGAgqB+Jg0r4d00i5N2JLZogWI2NRlF9N4MgIR8IwANrYfBuFLKCNgALobi2NJ5LRKL0QgjF5Cx3Qjqa5qJlYCwfGAuRwJoQiTkW0hMBoDBILgcgKJpJRoJg9Qbt0bFqYKXEic+pjGn0LndGJyhxMEYAuT6niiBkAAyIr5K0kylKI5R+GACChQ0YBNC0hQdL0yF9AM8AIJJExaNMABqWDwFAIV8Lk+EbBgWx8Ls+FHCcarnLkDDXLcYxid0K7YGuMHpWMcVlaCkndL26qGixEpSjKjbMs2Q60mpQJDbBPaqhS8JIvNFb0tKjIza5YwwFYDAisN2JrQaG0TXSkoMg2TYoh1fRRNlK1jKNwqiv+N1TXtD3tfNhKCn1q2EWNV3zU93TspyWDcphNVvQRaqfZtB2HN45D6iD70XeDIpo3N8qKmdH2vixOp6utBNtY981KY+XZIyNeP9k6GAsfA7qeiM76BiG4aRtGPpbWAinJqmYDpplWY5nmBaYMWZqPt9DCAXWf2zXTEL3p2Xo48jdoOuzLEjiwY4TlO47Jgic6Aouy7PN17xayS5tbnQO6mwer5HhgJ7+OefjJNI163rrKvPhqvN+vzX5wD+wuiwBNYa-drtuWBBwQXA0HYxCOGIWdqHoTyiO8YtSwsaNxGkYWFGi9RZ3dNJjHMaLtmqRxDkVxxS0wMJomixJzN9K3sLlR3YsJgzpB2d3GlaTpU76fIWBGUQpmLxZ6C8LgNkqexAI9xych4CstgDbkglOTS7na+1vSedigQ+X5AXRQoLWh40zRgHoxxjBwHaMoQugI96ECvkIUw-l0jgAAOrYHUpZPe84wC2yaBgZSq9kivk2GAHw2IbxkCwAiOgCgRBRR8DFMB-d-S2AACwAFYAC0AA2NhAAmAADGkT+sVQqgnoTYThABmFhoimGiNEWkUQ38wy5Q0HgMMUCwy2yoCAhafcBIwEIKfLG0wACcvDIGCNzDAtRUBfSeCAA)
```
{
    "Key": {
        "pk": "$taskapp#repositoryid_45-6620",
        "sk": "$repository#scores_90#linkid_23-3533#videolink_1"
                        `${entity_name}_${version}`-^^^^^^^^^^^                                  
    },
    "TableName": "your_table_name"
}
```
`get` a `link` with an `isolated` index: [link](https://electrodb.fun/?#code/PQKgBAsg9gJgpgGzARwK5wE4Es4GcA0YuccYGeqCALgUQBYCG5YA7llXWAGbZwB2MXGBDAAUKKwBbAA5QMVMAG8wAUT5V2AT0IBlTADcsAY1IBfbhiiSwAIkRwjVSzABGNgNzijUPrgVUGFwRSAF5bTShUDAB9AKC4aL4GSTgPcVAwABUGXABrITUNKk1hMW9fBXJZXHY5HCEwvjgWVXUtAApRMCUu7rBJWEQALh6+vv4izRGbACU4atqMTRt8XrH9TBqfaYBGFbW+4gxDE2nsvIBBaWkbA9NVsYYqJywXVCo8EcUD7qqoGqock0AEkYF8fn1itI4NM-Ng+ABzfZjMbkNBYchgsBOdAQ+4QorBcEoyGaaGwl6I5EkshwdGYkY4uB4h4o+C4IzYaQabajGlQmG2OFYKmsklo1AYuBYpksgkMBG4Yn8smCmzEKjUknsOCSJVCylIuUkzlwJ7Si5UZUkgXTPioSQuTBatlwLgMShWsDtACUYBCAD4wAARc0AOj4UBYvrFKOAwDARgYfEjCid-VgWC4OBgYAYXA+GET5HNMAhvzNMAA8nwEFNsRhcST8STUNIYKXLdaUbbbPbHc7Y2N46wnkZOFw5InGIi8NioHm+CUni83h9y6OqOPpiAXcOExrE1FyOpsVI8AEZKw6PwwG2Ox8yzSiHAvb7-UHQx8I1GYxuS9Wtb1rKzZ3LGIrwAAHp8fKovM-yLPW3zPt4CDBI4WC8jYfwAkCe59NIuTdiS2aIFiNiEfhYzeDICEfCMADa2HwbhSygjYAC6G4tjSeTESipEIOReRUX0I6muaiZWAsHxgLkcCaEIk5FtITAaAwSC4HICjySUaCYPUG7dDRMmCgxXHPqYxp9FZ3Q8cocTBGAVk+p4ogZAAMiK+StJMpSiOUfhgAg3kNGATQtIUHS9MhfQDPACD8RMWjTAAalg8BQF5fC5PhGwYFsfC7PhRwnGq5y5Aw1y3GMPHdCu2BrjBsVjCFOWgvx3S9uqhqiRKUoyo2zLNkOtIyUCHWwT2qoUvCSKjRW9LSoyQ3WWMMBWAwIqddiM0GnNfV0pKDINk2KJ1X0USJVNYzdcKor-kdA0rWdtWjYSgotdN5L7SK81rX07Kclg3KYUVN2kj9PUHaNF3dBycjNRud29bD70KvqX23Xt6qvqJOp6rNf01edo0SY+XYQ11OP9k6GCifA7qeiM76BiG4aRtGPoLWA4nJqmYDpvFWY5nmBaYMWZqPo9DCAXWL3DaTEL3p2XpY5Daq04OG4jiwY4TlO47Jgic6Aouy7PI17yKySetbnQO6iSOh5GMeExnikfjJNI163ir0vPhqLN+mzX5wD+XM8wBNby6dNs2WBBwQXA0GYxCOGITtqHoTy4PMeNSyiYRO3dIJ5GUTzxnSXRZn53RE0wJx3E83xVN9GXsK5ZXvMJuTpAmTXckKUpU6qfIWAaUQ2lD3p6C8LgRlSbRAK1wj5C4Csthtbk7EWTStlK7VvT2digROS5bmBQoVU+40zRgHoxzGHA7TKBngLz4Q29CKYrnpOAAB1bAsl9Lz3nGAE2TQMCSTHskV8mwwA+GxDeMgWAER0AUCIAKPggrvwbv6WwAAWAArAAWgAGzkIAEwAAY0hX2Ct5UEBCbBUIAMykLYcQthbC0iiBvmGZKGg8Bhm-mGE2VBX5jXrmxGAhA14wRsAATjoV-JhuZf7iKgL6dwQA)
```
{
    "Key": {
        "pk": "$taskapp#repositoryid_45-6620",
        "sk": "$repository#videolink_1#scores_90#linkid_23-3533"
                           ^^^^^^^^^^^-`${entity_name}_${version}`
    },
    "TableName": "your_table_name"
}
```

Where the `${entity_name}_${version}` is slotted is where the sort keys between entities will diverge. This is because it is unique break in the sorting key. The further these identifiers are to the right the more entities will mix with one another within a partition. The further these identifies are to the left the sooner the sort keys will diverge and entities will become isolated within a partition.

**Reply by Derek-X-Wang** [reply]
Thank you a lot for the detailed explanation! I understand now.
